be/src/storage/vectorized/merge_iterator.cpp

```diff
@@ -233,9 +261,116 @@ ChunkIteratorPtr new_merge_iterator(const std::vector<ChunkIteratorPtr>& childre
     for (size_t i = 0; i < children.size(); i += kMaxChildrenSize) {
         size_t j = std::min(i + kMaxChildrenSize, children.size());
         std::vector<ChunkIteratorPtr> v(children.begin() + i, children.begin() + j);
-        sub_merge_iterators.emplace_back(new_merge_iterator(v));
+        sub_merge_iterators.emplace_back(new_heap_merge_iterator(v));
+    }
+    return new_heap_merge_iterator(sub_merge_iterators);
+}
+
+// Merge iterator based on source masks.
+// The order of rows is determinate by mask sequence.
+class MaskMergeIterator final : public MergeIterator {
+public:
+    explicit MaskMergeIterator(std::vector<ChunkIteratorPtr> children, RowSourceMaskBuffer* mask_buffer)
+            : MergeIterator(std::move(children)), _chunks(_children.size()), _mask_buffer(mask_buffer) {
+        DCHECK(_mask_buffer);
+    }
+
+protected:
+    Status do_get_next(Chunk* chunk) override { return Status::NotSupported("get chunk not supported"); }
+    Status do_get_next(Chunk* chunk, std::vector<RowSourceMask>* source_masks) override;
+    Status fill(size_t child) override;
+
+private:
+    std::vector<MergeChunk> _chunks;
+    RowSourceMaskBuffer* _mask_buffer = nullptr;
+};
+
+inline Status MaskMergeIterator::do_get_next(Chunk* chunk, std::vector<RowSourceMask>* source_masks) {
+    if (!_inited) {
+        RETURN_IF_ERROR(init());
+    }
+    size_t rows = 0;
+    Status st;
+
+    auto st_or = _mask_buffer->has_remaining();
+    if (!st_or.ok()) {
+        return st_or.status();
+    }
+    while (st_or.value() && rows < _chunk_size) {
+        RowSourceMask mask = _mask_buffer->current();
+        uint16_t child = mask.get_source_num();
+        auto& min_chunk = _chunks[child];
+        DCHECK_GT(min_chunk.remaining_rows(), 0);
+
+        size_t offset = min_chunk.compared_row();
+        // check whether |min_chunk| has overlapping with others.
+        size_t min_chunk_num_rows = min_chunk._chunk->num_rows();
+        if (offset == 0 && _mask_buffer->has_same_source(child, min_chunk_num_rows)) {
+            if (rows == 0) {
+                chunk->swap_chunk(*min_chunk._chunk);
+                for (int i = 0; i < min_chunk_num_rows; ++i) {
+                    source_masks->emplace_back(_mask_buffer->current());
```

Comment:
Can you move this outside the loop?
```C++
source_masks->insert(source_masks->end(), min_chunk_num_rows, _mask_buffer->current())
```
As for the _maks_buffer, may be we can pass an adance steps to it:
```
_mask_buffer->advance(min_chunk_num_rows);
```